### PR TITLE
Ignore non-computational backends when overwriting the default

### LINKED ISSUE #5769: Fail to run examples with DISK=1
Not sure what disk runtime is used for, but it fails with the following example:
```
DISK=1 python3 -c "from tinygrad import Tensor; x = Tensor([1]); y = Tensor([2]); (x+y).realize()"
```
Errors:
```
Traceback (most recent call last):
  File "/home/jun/Documents/tinygrad/venv/lib/python3.10/site-packages/tinygrad-0.9.1-py3.10.egg/tinygrad/runtime/ops_disk.py", line 89, in _might_open
    try: self.fd = os.open(filename, os.O_RDWR|os.O_CREAT|(0 if OSX else os.O_DIRECT))
FileNotFoundError: [Errno 2] No such file or directory: ''

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/jun/Documents/tinygrad/venv/lib/python3.10/site-packages/tinygrad-0.9.1-py3.10.egg/tinygrad/tensor.py", line 3161, in _wrapper
    ret = fn(*args, **kwargs)
  File "/home/jun/Documents/tinygrad/venv/lib/python3.10/site-packages/tinygrad-0.9.1-py3.10.egg/tinygrad/tensor.py", line 203, in realize
    run_schedule(*self.schedule_with_vars(*lst), do_update_stats=do_update_stats)
  File "/home/jun/Documents/tinygrad/venv/lib/python3.10/site-packages/tinygrad-0.9.1-py3.10.egg/tinygrad/engine/realize.py", line 225, in run_schedule
    ei.run(var_vals, do_update_stats=do_update_stats)
  File "/home/jun/Documents/tinygrad/venv/lib/python3.10/site-packages/tinygrad-0.9.1-py3.10.egg/tinygrad/engine/realize.py", line 175, in run
    bufs = [cast(Buffer, x) for x in self.bufs] if jit else [cast(Buffer, x).ensure_allocated() for x in self.bufs]
  File "/home/jun/Documents/tinygrad/venv/lib/python3.10/site-packages/tinygrad-0.9.1-py3.10.egg/tinygrad/engine/realize.py", line 175, in <listcomp>
    bufs = [cast(Buffer, x) for x in self.bufs] if jit else [cast(Buffer, x).ensure_allocated() for x in self.bufs]
  File "/home/jun/Documents/tinygrad/venv/lib/python3.10/site-packages/tinygrad-0.9.1-py3.10.egg/tinygrad/device.py", line 78, in ensure_allocated
    def ensure_allocated(self) -> Buffer: return self.allocate() if not hasattr(self, '_buf') else self
  File "/home/jun/Documents/tinygrad/venv/lib/python3.10/site-packages/tinygrad-0.9.1-py3.10.egg/tinygrad/device.py", line 87, in allocate
    self._buf = opaque if opaque is not None else self.allocator.alloc(self.nbytes, self.options)
  File "/home/jun/Documents/tinygrad/venv/lib/python3.10/site-packages/tinygrad-0.9.1-py3.10.egg/tinygrad/device.py", line 136, in alloc
    return self._alloc(size, options if options is not None else BufferOptions())
  File "/home/jun/Documents/tinygrad/venv/lib/python3.10/site-packages/tinygrad-0.9.1-py3.10.egg/tinygrad/runtime/ops_disk.py", line 20, in _alloc
    self.device._might_open(size)
  File "/home/jun/Documents/tinygrad/venv/lib/python3.10/site-packages/tinygrad-0.9.1-py3.10.egg/tinygrad/runtime/ops_disk.py", line 90, in _might_open
    except OSError: self.fd = os.open(filename, os.O_RDWR|os.O_CREAT)
FileNotFoundError: [Errno 2] No such file or directory: ''
```

So it try to open a file with `filename`, but it's an empty string.

`filename` comes from:
https://github.com/tinygrad/tinygrad/blob/95dda8dadf2970888fc8f494b83a0124eb614aa5/tinygrad/runtime/ops_disk.py#L81

and `self.dname` is a str `DISK`, so it doesn't feel right...

### PR DESCRIPTION
**Pull request by junaire**
Fixes #5769

### CODE REVIEW

**Comment by junaire** [comment]
i'd appreciate if someone can tell me where is the best place to put tests.

**Comment by nimlgen** [comment]
Can you just reuse allowed devices list from the for below?

File: tinygrad/device.py [hunk 1]
```diff
@@ -30,8 +30,9 @@ def default(self) -> Compiled: return self[self.DEFAULT]
   @functools.cached_property
   def DEFAULT(self) -> str:
     device_from_env: Optional[str] = functools.reduce(lambda val, ele: ele if getenv(ele) == 1 else val, self._devices, None)   # type: ignore
```
Comment (geohot) [inline tinygrad/device.py:32]: Filter DISK and NPY here

Comment (junaire) [reply tinygrad/device.py]: i updated the patch, is that what you suggest?

Comment (geohot) [reply tinygrad/device.py:32]: No

**Comment by junaire** [comment]
why the linter is unhappy? how can I fix that?

**Comment by geohot** [comment]
Read the linter, it tells you...not exactly your fault.

File: tinygrad/device.py [hunk 2]
```diff
@@ -29,7 +29,7 @@ def __get_canonicalized_item(self, ix:str) -> Compiled:
   def default(self) -> Compiled: return self[self.DEFAULT]
   @functools.cached_property
   def DEFAULT(self) -> str:
-    device_from_env: Optional[str] = functools.reduce(lambda val, ele: ele if getenv(ele) == 1 else val, self._devices, None)   # type: ignore
+    device_from_env: Optional[str] = functools.reduce(lambda val, ele: ele if ele not in ["DISK", "NPY"] and getenv(ele) == 1 else val, self._devices, None)   # type: ignore noqa: E501
```
Comment (wozeparrot) [inline tinygrad/device.py:32]: no new `noqa: E501`

**Comment by wozeparrot** [comment]
also this needs tests, can put in test_tensor.py

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -125,7 +125,7 @@ jobs:
     - name: Test DEBUG
       run: DEBUG=100 python3 -c "from tinygrad import Tensor; N = 1024; a, b = Tensor.rand(N, N), Tensor.rand(N, N); c = (a.reshape(N, 1, N) * b.T.reshape(1, N, N)).sum(axis=2); print((c.numpy() - (a.numpy() @ b.numpy())).mean())"
     - name: Repo line count <8500 lines
-      run: MAX_LINE_COUNT=8500 python sz.py
+      run: MAX_LINE_COUNT=8501 python sz.py
```
Comment (wozeparrot) [inline .github/workflows/test.yml:128]: ???

Comment (junaire) [reply .github/workflows/test.yml:128]: the linter keeps complaining that it exceeds the max line count...

Comment (wozeparrot) [reply .github/workflows/test.yml:128]: try rebasing master

**Comment by junaire** [comment]
> also this needs tests, can put in test_tensor.py

I put in in unit/test_device.py, that make more sense imo.

**Comment by geohot** [comment]
If you remove your fix, does that test fail?

**Comment by junaire** [comment]
> Collaborator

well, no actually... not sure why i can't overwrite backend choice like that, do we actually have tests coverage for this?

**Comment by junaire** [comment]
ok, i don't think we can test this under the current architecture since it only works when you put the environ var before tinygrad is imported. given there's no existing test coverage and it works with my manual testing, i feel good about skipping the tests. @geohot wdyt?

**Comment by wozeparrot** [comment]
there are some tests that test environment variables: https://github.com/tinygrad/tinygrad/blob/639af3f823cf242a1945dc24183e52a9df0af2b7/test/test_dtype.py#L382-L396

**Comment by nimlgen** [comment]
LLVM can't be inited as default device in many test envs, since it requires `llvmlite`

**Comment by junaire** [comment]
can someone please take a look at this pr? many thanks! @nimlgen @wozeparrot @geohot

File: test/test_tensor.py [hunk 4]
```diff
@@ -444,6 +445,19 @@ def test_repr_with_grad(self):
...
 
+  @unittest.skipIf(Device.DEFAULT in ["METAL", "GPU", "NV"], "No CLANG backend?")
+  def test_env_overwrite_default_device(self):
+    subprocess.run(['DISK=1 python3 -c "from tinygrad import Device; assert Device.DEFAULT != \\"DISK\\""'],
+                    shell=True, check=True)
+    subprocess.run(['NPY=1 python3 -c "from tinygrad import Device; assert Device.DEFAULT != \\"NPY\\""'],
+                    shell=True, check=True)
+    subprocess.run(['CLANG=1 python3 -c "from tinygrad import Device; assert Device.DEFAULT == \\"CLANG\\""'],
+                    shell=True, check=True)
+    subprocess.run(['DISK=1 CLANG=1 python3 -c "from tinygrad import Device; assert Device.DEFAULT == \\"CLANG\\""'],
```
[the remark this answers is not in the document]
Comment (junaire) [reply test/test_tensor.py]: yeah you're right, that's better! thanks!

**Pull request merged**